### DISCUSSION TITLE
(novice) select one of two stimuli randomly with replacement, repeatedly

### DISCUSSION THREAD

**Discussion by preinagel** [bonsai-rx/bonsai#1520] [Q&A] [answered]
[sorry if this has been addressed before, I did search, but didn't find it]
As a beginner exercise, I am trying to implement a basic workflow where a keypress makes a grating appear for 1s, selecting either parameter 1 or 2 at random each trial.  I managed to create a workflow that will present the stimuli in random order once and then stop. (I'm using Permutation and there are only 2 stimulus values but for some reason I get three trials; the first trial seems to choose a parameter from the distribution of the previous run, but whatever). 

**Where I am stuck is** how to continue randomly selecting one of the two stimuli at random, indefinitely.  Here's what I've got so far:

![image](https://github.com/bonsai-rx/bonsai/assets/80356291/519f5ed1-91e9-4ca3-9a2e-25180a21de45)

Where "GratingAtLoc" contains:
![image](https://github.com/bonsai-rx/bonsai/assets/80356291/dac03366-f9fe-4d35-b48e-2aee9564d30a)

and "Stim1" contains:
![image](https://github.com/bonsai-rx/bonsai/assets/80356291/d9ff5572-516f-4795-b192-8b1add0d6ea0)

It seemed like Repeat or RepeatCount would work, but I can't seem to figure out where/how to connect them in the workflow.

Also, I don't want to repeat in a way that every pair of trials is a permutation of the two values, I just want the whole sequence to be i.i.d.  So maybe a different random distribution operator would be better than Permutation?

I can see I'm going to have a ton of questions of this kind, which doesn't seem appropriate for a forum. Is there some other training resource I should be tapping?

**Comment by bruno-f-cruz** [comment, marked as answer]
As any other programming language, there are likely an infinite number of ways to achieve what you are asking. I have coded an example where you define a sort of "block" structure at the top that gets permutated on each block. When the block ends, repeat retriggers the whole pipeline.

[1520.zip](https://github.com/bonsai-rx/bonsai/files/12359587/1520.zip)

PS: In the future include your workflow as suggested, I could have edited your original workflow directly :)

**Reply by preinagel** [reply]
Thanks, and sorry I didn't realize I could attach the workflow file, will do in the future.

**Reply by preinagel** [reply]
BTW, I just found the answer to "Is there some other training resource I should be tapping?" which is: the NeuroGears course on YouTube https://youtube.com/playlist?list=PL8wJDYCky9yiNucwLUgvbntawVpodLSGf&si=Ld72nc9MnJmR8uiI 

That course covers the ground-zero fundamentals I was missing to be able to interpret the rest of the available documentation and online resources. Pretty much every other question I have asked on the forum was due to lacking that information.

Noting this here in case another newcomer searches for this question and pulls up this thread.
